File: mobile/ios/RoadtripCopilot/Services/GooglePlacesAPIClient.swift
//
//  GooglePlacesAPIClient.swift
//  RoadtripCopilot
//
//  Google Places API client for POI discovery failover
//  Performance target: <1000ms response time
//

import Foundation
import CoreLocation
import os.log

/// Google Places API client for POI discovery with automotive safety compliance
@available(iOS 15.0, *)
class GooglePlacesAPIClient: ObservableObject {
    
    // MARK: - Properties
    static let shared = GooglePlacesAPIClient()
    
    @Published var isProcessing = false
    @Published var lastError: Error?
    
    private let apiKey = "YOUR_GOOGLE_PLACES_API_KEY" // Replace with actual API key
    private let baseURL = "https://maps.googleapis.com/maps/api/place"
    private let session: URLSession
    private let logger = Logger(subsystem: "com.hmi2.roadtrip-copilot", category: "GooglePlacesAPI")
    
    // Performance monitoring
    private var requestStartTime: Date?
    private let performanceThreshold: TimeInterval = 1.0 // 1000ms target
    
    // Rate limiting
    private var lastRequestTime: Date?
    private let minimumRequestInterval: TimeInterval = 0.1 // 100ms between requests
    
    private init() {
        // Configure session for automotive use (timeout settings)
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 8.0 // 8 second timeout
        config.timeoutIntervalForResource = 10.0 // 10 second resource timeout
        config.waitsForConnectivity = false // Don't wait for connectivity
        self.session = URLSession(configuration: config)
    }
    
    // MARK: - Public API
    
    /// Search for POIs near a location with automotive safety optimizations
    /// - Parameters:
    ///   - location: Center point for search
    ///   - category: POI category to search for
    ///   - radius: Search radius in meters (max 50km for safety)
    ///   - maxResults: Maximum number of results (limited to 20 for automotive safety)
    /// - Returns: Array of discovered POIs
    func searchPOIs(
        near location: CLLocation,
        category: String,
        radius: Double = 5000,
        maxResults: Int = 10
    ) async throws -> [POIData] {
        
        logger.info("🔍 [PLACES API] Starting POI search near \(location.coordinate.latitude), \(location.coordinate.longitude)")
        requestStartTime = Date()
        
        // Rate limiting check
        if let lastTime = lastRequestTime,
           Date().timeIntervalSince(lastTime) < minimumRequestInterval {
            try await Task.sleep(nanoseconds: UInt64(minimumRequestInterval * 1_000_000_000))
        }
        lastRequestTime = Date()
        
        isProcessing = true
        defer { 
            isProcessing = false 
            logPerformance()
        }
        
        // Validate inputs for automotive safety
        let safeRadius = min(radius, 50000) // Max 50km for safety
        let safeMaxResults = min(maxResults, 20) // Max 20 results for automotive UI
        
        // Convert category to Google Places type
        let placeType = mapCategoryToGoogleType(category)
        
        // Build request URL
        let endpoint = "\(baseURL)/nearbysearch/json"
        var components = URLComponents(string: endpoint)!
        
        components.queryItems = [
            URLQueryItem(name: "location", value: "\(location.coordinate.latitude),\(location.coordinate.longitude)"),
            URLQueryItem(name: "radius", value: "\(Int(safeRadius))"),
            URLQueryItem(name: "type", value: placeType),
            URLQueryItem(name: "key", value: apiKey)
        ]
        
        guard let url = components.url else {
            throw PlacesAPIError.invalidURL
        }
        
        do {
            // Make request with performance monitoring
            let (data, response) = try await session.data(from: url)
            
            // Validate response
            guard let httpResponse = response as? HTTPURLResponse else {
                throw PlacesAPIError.invalidResponse
            }
            
            guard httpResponse.statusCode == 200 else {
                logger.error("❌ [PLACES API] HTTP error: \(httpResponse.statusCode)")
                throw PlacesAPIError.httpError(statusCode: httpResponse.statusCode)
            }
            
            // Parse response
            let placesResponse = try JSONDecoder().decode(GooglePlacesResponse.self, from: data)
            
            // Check API status
            guard placesResponse.status == "OK" || placesResponse.status == "ZERO_RESULTS" else {
                logger.error("❌ [PLACES API] API error: \(placesResponse.status)")
                throw PlacesAPIError.apiError(status: placesResponse.status, message: placesResponse.errorMessage)
            }
            
            // Convert to POIData
            let pois = placesResponse.results.prefix(safeMaxResults).compactMap { place in
                convertPlaceToPOI(place, originalLocation: location, category: category)
            }
            
            logger.info("✅ [PLACES API] Found \(pois.count) POIs for category '\(category)'")
            return pois
            
        } catch {
            logger.error("❌ [PLACES API] Request failed: \(error.localizedDescription)")
            lastError = error
            throw error
        }
    }
    
    /// Get detailed information about a specific POI
    func getPOIDetails(placeId: String) async throws -> PlaceDetails {
        logger.info("📋 [PLACES API] Getting details for place ID: \(placeId)")
        
        let endpoint = "\(baseURL)/details/json"
        var components = URLComponents(string: endpoint)!
        
        components.queryItems = [
            URLQueryItem(name: "place_id", value: placeId),
            URLQueryItem(name: "fields", value: "name,rating,formatted_phone_number,opening_hours,website,reviews,photos"),
            URLQueryItem(name: "key", value: apiKey)
        ]
        
        guard let url = components.url else {
            throw PlacesAPIError.invalidURL
        }
        
        let (data, _) = try await session.data(from: url)
        let response = try JSONDecoder().decode(PlaceDetailsResponse.self, from: data)
        
        guard response.status == "OK" else {
            throw PlacesAPIError.apiError(status: response.status, message: response.errorMessage)
        }
        
        return response.result
    }
    
    // MARK: - Helper Methods
    
    private func mapCategoryToGoogleType(_ category: String) -> String {
        // Map POI categories to Google Places types
        let mapping: [String: String] = [
            "restaurant": "restaurant",
            "food": "restaurant",
            "gas_station": "gas_station",
            "lodging": "lodging",
            "hotel": "lodging",
            "attraction": "tourist_attraction",
            "tourist_attraction": "tourist_attraction",
            "park": "park",
            "museum": "museum",
            "shopping": "shopping_mall",
            "hospital": "hospital",
            "pharmacy": "pharmacy",
            "bank": "bank",
            "atm": "atm",
            "church": "church",
            "school": "school"
        ]
        
        return mapping[category.lowercased()] ?? "point_of_interest"
    }
    
    private func convertPlaceToPOI(_ place: GooglePlace, originalLocation: CLLocation, category: String) -> POIData? {
        guard let lat = place.geometry?.location.lat,
              let lng = place.geometry?.location.lng else {
            return nil
        }
        
        let poiLocation = CLLocation(latitude: lat, longitude: lng)
        let distance = originalLocation.distance(from: poiLocation) / 1000.0 // Convert to km
        
        return POIData(
            name: place.name,
            location: poiLocation,
            category: category,
            rating: place.rating ?? 0.0,
            imageURL: place.photos?.first?.photoReference.flatMap { photoRef in
                URL(string: "\(baseURL)/photo?maxwidth=400&photoreference=\(photoRef)&key=\(apiKey)")
            },
            reviewSummary: place.reviews?.first?.text,
            distanceFromUser: distance,
            couldEarnRevenue: place.rating ?? 0.0 >= 4.0 // High-rated places more likely to earn revenue
        )
    }
    
    private func logPerformance() {
        guard let startTime = requestStartTime else { return }
        let elapsed = Date().timeIntervalSince(startTime)
        
        if elapsed > performanceThreshold {
            logger.warning("⚠️ [PLACES API] Slow request: \(String(format: "%.0f", elapsed * 1000))ms (target: \(Int(performanceThreshold * 1000))ms)")
        } else {
            logger.info("✅ [PLACES API] Request completed in \(String(format: "%.0f", elapsed * 1000))ms")
        }
    }
}

// MARK: - Data Models

struct GooglePlacesResponse: Codable {
    let results: [GooglePlace]
    let status: String
    let errorMessage: String?
    
    enum CodingKeys: String, CodingKey {
        case results, status
        case errorMessage = "error_message"
    }
}

struct GooglePlace: Codable {
    let placeId: String
    let name: String
    let rating: Double?
    let priceLevel: Int?
    let vicinity: String?
    let geometry: PlaceGeometry?
    let photos: [PlacePhoto]?
    let reviews: [PlaceReview]?
    
    enum CodingKeys: String, CodingKey {
        case name, rating, vicinity, geometry, photos, reviews
        case placeId = "place_id"
        case priceLevel = "price_level"
    }
}

struct PlaceGeometry: Codable {
    let location: PlaceLocation
}

struct PlaceLocation: Codable {
    let lat: Double
    let lng: Double
}

struct PlacePhoto: Codable {
    let photoReference: String
    let height: Int
    let width: Int
    
    enum CodingKeys: String, CodingKey {
        case height, width
        case photoReference = "photo_reference"
    }
}

struct PlaceReview: Codable {
    let text: String
    let rating: Int
    let authorName: String
    let time: Int
    
    enum CodingKeys: String, CodingKey {
        case text, rating, time
        case authorName = "author_name"
    }
}

// Place Details Models
struct PlaceDetailsResponse: Codable {
    let result: PlaceDetails
    let status: String
    let errorMessage: String?
    
    enum CodingKeys: String, CodingKey {
        case result, status
        case errorMessage = "error_message"
    }
}

struct PlaceDetails: Codable {
    let name: String
    let rating: Double?
    let formattedPhoneNumber: String?
    let website: String?
    let openingHours: OpeningHours?
    let reviews: [PlaceReview]?
    let photos: [PlacePhoto]?
    
    enum CodingKeys: String, CodingKey {
        case name, rating, website, reviews, photos
        case formattedPhoneNumber = "formatted_phone_number"
        case openingHours = "opening_hours"
    }
}

struct OpeningHours: Codable {
    let openNow: Bool
    let weekdayText: [String]?
    
    enum CodingKeys: String, CodingKey {
        case weekdayText = "weekday_text"
        case openNow = "open_now"
    }
}

// MARK: - Error Handling

enum PlacesAPIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case httpError(statusCode: Int)
    case apiError(status: String, message: String?)
    case rateLimitExceeded
    case noResults
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid Google Places API URL"
        case .invalidResponse:
            return "Invalid response from Google Places API"
        case .httpError(let statusCode):
            return "HTTP error: \(statusCode)"
        case .apiError(let status, let message):
            return "Google Places API error (\(status)): \(message ?? "Unknown error")"
        case .rateLimitExceeded:
            return "Google Places API rate limit exceeded"
        case .noResults:
            return "No POIs found for the specified criteria"
        }
    }
}

File: mobile/android/app/src/main/java/com/roadtrip/copilot/services/GooglePlacesAPIClient.kt
package com.roadtrip.copilot.services

import android.util.Log
import com.roadtrip.copilot.models.POIData
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.delay
import org.json.JSONObject
import java.net.URL
import java.net.URLEncoder
import java.text.DecimalFormat
import kotlin.math.*

/**
 * Google Places API client for POI discovery with automotive safety compliance
 * Performance target: <1000ms response time
 * Equivalent to iOS GooglePlacesAPIClient.swift implementation
 */
class GooglePlacesAPIClient {
    
    companion object {
        private const val TAG = "GooglePlacesAPI"
        val shared = GooglePlacesAPIClient()
        
        // Performance monitoring
        private const val PERFORMANCE_THRESHOLD_MS = 1000L // 1000ms target
        private const val MINIMUM_REQUEST_INTERVAL_MS = 100L // 100ms between requests
        
        // Automotive safety limits
        private const val MAX_RADIUS_METERS = 50000 // Max 50km for safety
        private const val MAX_RESULTS = 20 // Max 20 results for automotive UI
    }
    
    private val apiKey = "YOUR_GOOGLE_PLACES_API_KEY" // Replace with actual API key
    private val baseURL = "https://maps.googleapis.com/maps/api/place"
    
    var isProcessing = false
        private set
    var lastError: Exception? = null
        private set
    
    // Rate limiting
    private var lastRequestTime: Long = 0
    
    /**
     * Search for POIs near a location with automotive safety optimizations
     * @param latitude Center point latitude for search
     * @param longitude Center point longitude for search
     * @param category POI category to search for
     * @param radiusMeters Search radius in meters (max 50km for safety)
     * @param maxResults Maximum number of results (limited to 20 for automotive safety)
     * @return List of discovered POIs
     */
    suspend fun searchPOIs(
        latitude: Double,
        longitude: Double,
        category: String,
        radiusMeters: Double = 5000.0,
        maxResults: Int = 10
    ): List<POIData> = withContext(Dispatchers.IO) {
        
        Log.i(TAG, "🔍 [PLACES API] Starting POI search near $latitude, $longitude")
        val requestStartTime = System.currentTimeMillis()
        
        try {
            // Rate limiting check
            val currentTime = System.currentTimeMillis()
            if (currentTime - lastRequestTime < MINIMUM_REQUEST_INTERVAL_MS) {
                delay(MINIMUM_REQUEST_INTERVAL_MS - (currentTime - lastRequestTime))
            }
            lastRequestTime = System.currentTimeMillis()
            
            isProcessing = true
            
            // Validate inputs for automotive safety
            val safeRadius = min(radiusMeters, MAX_RADIUS_METERS.toDouble())
            val safeMaxResults = min(maxResults, MAX_RESULTS)
            
            // Convert category to Google Places type
            val placeType = mapCategoryToGoogleType(category)
            
            // Build request URL
            val location = "$latitude,$longitude"
            val encodedLocation = URLEncoder.encode(location, "UTF-8")
            val encodedType = URLEncoder.encode(placeType, "UTF-8")
            
            val endpoint = "$baseURL/nearbysearch/json" +
                    "?location=$encodedLocation" +
                    "&radius=${safeRadius.toInt()}" +
                    "&type=$encodedType" +
                    "&key=$apiKey"
            
            // Make request with performance monitoring
            val url = URL(endpoint)
            val connection = url.openConnection()
            connection.setRequestProperty("Accept", "application/json")
            connection.connectTimeout = 8000 // 8 second timeout
            connection.readTimeout = 10000 // 10 second read timeout
            
            val response = connection.getInputStream().bufferedReader().use { it.readText() }
            
            // Parse response
            val jsonResponse = JSONObject(response)
            val status = jsonResponse.optString("status")
            
            if (status != "OK" && status != "ZERO_RESULTS") {
                val errorMessage = jsonResponse.optString("error_message", "Unknown error")
                Log.e(TAG, "❌ [PLACES API] API error: $status - $errorMessage")
                throw Exception("Google Places API error ($status): $errorMessage")
            }
            
            // Convert results to POIData
            val results = jsonResponse.optJSONArray("results") ?: return@withContext emptyList()
            val pois = mutableListOf<POIData>()
            
            for (i in 0 until minOf(results.length(), safeMaxResults)) {
                val place = results.getJSONObject(i)
                val poi = convertPlaceToPOI(place, latitude, longitude, category)
                if (poi != null) {
                    pois.add(poi)
                }
            }
            
            // Log performance
            val elapsed = System.currentTimeMillis() - requestStartTime
            if (elapsed > PERFORMANCE_THRESHOLD_MS) {
                Log.w(TAG, "⚠️ [PLACES API] Slow request: ${elapsed}ms (target: ${PERFORMANCE_THRESHOLD_MS}ms)")
            } else {
                Log.i(TAG, "✅ [PLACES API] Request completed in ${elapsed}ms")
            }
            
            Log.i(TAG, "✅ [PLACES API] Found ${pois.size} POIs for category '$category'")
            return@withContext pois
            
        } catch (e: Exception) {
            Log.e(TAG, "❌ [PLACES API] Request failed: ${e.message}", e)
            lastError = e
            throw e
        } finally {
            isProcessing = false
        }
    }
    
    /**
     * Get detailed information about a specific POI
     */
    suspend fun getPOIDetails(placeId: String): PlaceDetails = withContext(Dispatchers.IO) {
        Log.i(TAG, "📋 [PLACES API] Getting details for place ID: $placeId")
        
        val fields = "name,rating,formatted_phone_number,opening_hours,website,reviews,photos"
        val encodedPlaceId = URLEncoder.encode(placeId, "UTF-8")
        val encodedFields = URLEncoder.encode(fields, "UTF-8")
        
        val endpoint = "$baseURL/details/json" +
                "?place_id=$encodedPlaceId" +
                "&fields=$encodedFields" +
                "&key=$apiKey"
        
        val url = URL(endpoint)
        val connection = url.openConnection()
        connection.setRequestProperty("Accept", "application/json")
        
        val response = connection.getInputStream().bufferedReader().use { it.readText() }
        val jsonResponse = JSONObject(response)
        
        val status = jsonResponse.optString("status")
        if (status != "OK") {
            val errorMessage = jsonResponse.optString("error_message", "Unknown error")
            throw Exception("Google Places API error ($status): $errorMessage")
        }
        
        val result = jsonResponse.getJSONObject("result")
        return@withContext PlaceDetails(
            name = result.optString("name", ""),
            rating = result.optDouble("rating", 0.0),
            phoneNumber = result.optString("formatted_phone_number"),
            website = result.optString("website"),
            isOpenNow = result.optJSONObject("opening_hours")?.optBoolean("open_now"),
            weekdayText = result.optJSONObject("opening_hours")
                ?.optJSONArray("weekday_text")
                ?.let { array ->
                    (0 until array.length()).map { array.getString(it) }
                },
            reviews = result.optJSONArray("reviews")?.let { array ->
                (0 until minOf(array.length(), 3)).map { index ->
                    val review = array.getJSONObject(index)
                    PlaceReview(
                        text = review.optString("text", ""),
                        rating = review.optInt("rating", 0),
                        authorName = review.optString("author_name", ""),
                        time = review.optLong("time", 0)
                    )
                }
            } ?: emptyList(),
            photos = result.optJSONArray("photos")?.let { array ->
                (0 until minOf(array.length(), 3)).map { index ->
                    val photo = array.getJSONObject(index)
                    PlacePhoto(
                        photoReference = photo.optString("photo_reference", ""),
                        height = photo.optInt("height", 0),
                        width = photo.optInt("width", 0)
                    )
                }
            } ?: emptyList()
        )
    }
    
    // MARK: - Helper Methods
    
    private fun mapCategoryToGoogleType(category: String): String {
        // Map POI categories to Google Places types
        return when (category.lowercase()) {
            "restaurant", "food" -> "restaurant"
            "gas_station" -> "gas_station"
            "lodging", "hotel" -> "lodging"
            "attraction", "tourist_attraction" -> "tourist_attraction"
            "park" -> "park"
            "museum" -> "museum"
            "shopping" -> "shopping_mall"
            "hospital" -> "hospital"
            "pharmacy" -> "pharmacy"
            "bank" -> "bank"
            "atm" -> "atm"
            "church" -> "church"
            "school" -> "school"
            else -> "point_of_interest"
        }
    }
    
    private fun convertPlaceToPOI(
        place: JSONObject, 
        searchLat: Double, 
        searchLng: Double, 
        category: String
    ): POIData? {
        val geometry = place.optJSONObject("geometry") ?: return null
        val location = geometry.optJSONObject("location") ?: return null
        
        val lat = location.optDouble("lat", 0.0)
        val lng = location.optDouble("lng", 0.0)
        
        if (lat == 0.0 && lng == 0.0) return null
        
        // Calculate distance using Haversine formula
        val distance = calculateDistance(searchLat, searchLng, lat, lng)
        
        val name = place.optString("name", "Unknown Place")
        val rating = place.optDouble("rating", 0.0)
        val vicinity = place.optString("vicinity", "")
        val placeId = place.optString("place_id", "")
        
        // Generate image URL from photos
        val photos = place.optJSONArray("photos")
        val imageURL = if (photos != null && photos.length() > 0) {
            val photo = photos.getJSONObject(0)
            val photoReference = photo.optString("photo_reference")
            if (photoReference.isNotEmpty()) {
                "$baseURL/photo?maxwidth=400&photoreference=$photoReference&key=$apiKey"
            } else null
        } else null
        
        // Get review summary
        val reviews = place.optJSONArray("reviews")
        val reviewSummary = if (reviews != null && reviews.length() > 0) {
            reviews.getJSONObject(0).optString("text", null)
        } else null
        
        return POIData(
            id = placeId.ifEmpty { generatePOIId(name, lat, lng) },
            name = name,
            description = vicinity.ifEmpty { "A point of interest near your location" },
            category = category,
            latitude = lat,
            longitude = lng,
            distanceFromUser = distance,
            rating = rating,
            imageURL = imageURL,
            reviewSummary = reviewSummary,
            couldEarnRevenue = rating >= 4.0, // High-rated places more likely to earn revenue
            address = vicinity.ifEmpty { null },
            priceLevel = place.optInt("price_level", 2)
        )
    }
    
    private fun calculateDistance(lat1: Double, lng1: Double, lat2: Double, lng2: Double): Double {
        val earthRadius = 6371.0 // Earth radius in kilometers
        
        val dLat = Math.toRadians(lat2 - lat1)
        val dLng = Math.toRadians(lng2 - lng1)
        
        val a = sin(dLat / 2) * sin(dLat / 2) +
                cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) *
                sin(dLng / 2) * sin(dLng / 2)
        
        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
        
        // Convert to miles for consistency with iOS implementation
        return earthRadius * c * 0.621371
    }
    
    private fun generatePOIId(name: String, lat: Double, lng: Double): String {
        val formatter = DecimalFormat("#.####")
        val latStr = formatter.format(lat)
        val lngStr = formatter.format(lng)
        return "${name.hashCode()}_${latStr}_${lngStr}"
    }
}

// MARK: - Data Models

data class PlaceDetails(
    val name: String,
    val rating: Double,
    val phoneNumber: String?,
    val website: String?,
    val isOpenNow: Boolean?,
    val weekdayText: List<String>?,
    val reviews: List<PlaceReview>,
    val photos: List<PlacePhoto>
)

data class PlaceReview(
    val text: String,
    val rating: Int,
    val authorName: String,
    val time: Long
)

data class PlacePhoto(
    val photoReference: String,
    val height: Int,
    val width: Int
)